### PR TITLE
Fix potential infinite loop

### PR DESCRIPTION
**Pull request by sbird**
This seems to fix the hang we were seeing in MP-Gadget.

### CODE REVIEW

**Comment by rainwoodman** [comment]
Quick comment.

This part of code was tricky.

Why do you need to change Pmin to -1? The old value may have to do with the
close-open ness of the pivot points -- breaking that invariance may have
unintended consequences.

Also do you have a scebario to illustrate the dead loop? We can add a test
case for it in the python layer.

**Comment by sbird** [comment]
It is the other way around: I changed Pmin to 0 from -1. It only takes this value if the number of elements on this processor is zero, in which case we want Pmax == Pmin because there is nothing to bisect around. And the bisect bracket is documented to be both closed, so unless the documentation is lying this is the right thing.

It seems that because the radix is really an unsigned type of at least 8 bytes, the -1, -1, -1, -1 bytefield was being interpreted oddly, and I think overflowing in some strange (I think compiler-dependent) way, so that, eg, the FOF radix sometimes stored -2, -1, -1, -1 if the radix was 24 bytes.

The test case is in the MP-Gadget source, test_mpsort.c with the very large radix.

**Comment by sbird** [comment]
Although I might be wrong here. There are still odd crashes happening.

**Comment by rainwoodman** [comment]
Radix is unsigned, so 0xfffffffff... would be an extremely large number
such that the closed bracket is empty

but [0,0] is not empty. That could be the reason I was using [-1, 0].

At certain placed I may had to +1 to the pivots. If something went wrong
there? E.g. forgot to cast to unsigned?

**Comment by sbird** [comment]
This only seems to manifest on the large (>u8) radices, and seems to be a single entry -1 from what it should be, so I wondered if it was something to do with the carry bit?

**Comment by sbird** [comment]
What do you think about this? I still suspect a bug in the radix function, but this makes it benign, and seems to make sense: it marks the empty set as narrow!

**Comment by sbird** [comment]
It does fix the MP-Gadget examples

**Comment by sbird** [comment]
The build failure is just a download timeout

File: internal-parallel.h [hunk 1]
```diff
@@ -184,7 +184,7 @@ static void piter_bisect(struct piter * pi, char * P) {
              * the range is too small, so we set flag narrow, 
              * and next iteration we will directly test Pright */
             if(d->compar(&P[i * d->rsize], 
-                &pi->Pleft[i * d->rsize], d->rsize) == 0) {
+                &pi->Pleft[i * d->rsize], d->rsize) <= 0) {
```
Comment (rainwoodman) [inline internal-parallel.h]: This looks like a reasonable cure, but we are not at the bottom of it. Something must have been seriously wrong if after bisect P is not inside [left right].

I am thinking of adding an assertion before and after 180, die and print everything if the precondition that P is in side [left, right] is wrong.

If you have other priorities, I can take a closer look into this over the weekend.

File: radixsort.c [hunk 2]
```diff
@@ -114,6 +114,10 @@ static void _bisect_radix(void * r, const void * r1, const void * r2, size_t rsi
     const unsigned char * u2 = r2;
     unsigned char * u = r;
     unsigned int carry = 0;
+    if(dir > 0) {
+        u1 += rsize - 1;
+        u2 += rsize - 1;
+    }
```
Comment (rainwoodman) [inline radixsort.c]: This didn't bite us, because it only affected be, but the tests (and use cases) only covered le?

Comment (sbird) [reply radixsort.c]: Right!

**Comment by sbird** [comment]
Something is still wrong: the tests now pass but when I run Gadget master the hydro test fails with an error (on 12 MPIs and 2 OMP threads: it is fine at lower parallel) as soon as it gets to the first non-trivial linking step. The error is:

[ 001005.09 ] Comoving linking length: 12.5
[ 001005.09 ] Start linking particles (presently allocated=0.0820312 MB)
[ 001005.29 ] Linked 3903 particles 0.190087 seconds
[ 001005.48 ] Linked 0 particles 0.181972 seconds
[ 001005.48 ] Local groups found.
[ 001005.48 ] Group finding done.
[ 001005.48 ] Start finding nearest dm-particle (presently allocated=0.0820312 MB)
[ 001005.48 ] fof-nearest iteration started
[ 001005.89 ] fof-nearest iteration 0: need to repeat for 0000000002 /0000262146 particles.
[ 001006.30 ] fof-nearest iteration 1: need to repeat for 0000000000 /0000262146 particles.
[ 001006.30 ] Attached gas and star particles to nearest dm particles.
[ 001006.32 ] Compiled local group data and catalogue.
[ 001006.33 ] Task 6: i=0 Group base Length 65 != Group Length 0

which I think may indicate that mpsort_mpi is not sorting correctly (but I need to check this).

When the build flags honoring code went in I updated from bdb915ef63d511222efefb527e09c3dcebc4c6d4 to  fbfe28726e16346d93d968a4 , which includes the code to gather into segments: this seems like a good place to start looking to me. 

I do unfortunately have other things I need to take care of, and I don't really understand the MP-Sort code still, so I would very much appreciate it if you could take a look this weekend!

**Comment by sbird** [comment]
See PR https://github.com/MP-Gadget/MP-Gadget/pull/361 for a test which checks that the output of MP-Sort is actually sorted (and fails). Run it on hydro with >= 12 MPIs and it will fail at z ~ 12.

**Comment by rainwoodman** [comment]
Only known failure is on stampede?

It is hard to picture a bug this bad in MPSort lurking for so long not
affected runs (fastpm also use it to sort halos and particles).

It may go away automatically on a different platform.I will do that the
first thing on Cori on Saturday.

Also probably a good idea to make sure I can access the paths for your
example job script and build files on stampede.

**Comment by sbird** [comment]
No, it happens on my workstation, with openmpi and gcc 9 (and optimizations off). Address sanitizer is clean but it only happens with a low number of particles per rank. I can make the problem disappear by doing mpsort_mpi_set_options(MPSORT_DISABLE_GATHER_SORT); in fof_init.

**Comment by sbird** [comment]
The example job script is just the hydro example. This is my Options.mk:

#Uncomment below to specify default options
MPICC       =   mpicc
MPICXX       =   mpic++

# For GCC add -mpc64 if IEEE 64bit FP is desired.
OPTIMIZE =  -fopenmp -O0 -std=gnu99 -g -Wall -Wextra -Wno-unused-parameter -Wvla -fsanitize=address
GSL_INCL = $(shell pkg-config --cflags gsl)
GSL_LIBS = $(shell pkg-config --libs gsl)

#--------------------------------------- Basic operation mode of code
#OPT += -DLIGHTCONE                       # write a lightcone on the fly; in development
OPT += -DVALGRIND     # allow debugging with valgrind, disable the GADGET memory allocator.
OPT += -DDEBUG      # print a lot of debugging messages

% mpicc --showme
gcc -pthread -Wl,-rpath -Wl,/usr/lib/openmpi -Wl,--enable-new-dtags -L/usr/lib/openmpi -lmpi
% mpicc --version
gcc (GCC) 9.2.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Comment by sbird** [comment]
Using
OMP_NUM_THREADS=2 mpirun -np 12 ../../gadget/MP-Gadget paramfile.gadget
in examples/hydro

**Comment by sbird** [comment]
My guess is that usually there are enough objects per rank that the buggy code does not run. The sort failure happens with ~ 2 halos per rank.

**Comment by rainwoodman** [comment]
It takes a very long time to get there on my desktop computer....

**Comment by sbird** [comment]
My apologies. Cori might be better...I think it needed at least 8 MPIs to trigger

**Comment by rainwoodman** [comment]
It ran and crashed.

I found a likely reproducer with the python tests, and filed a fix.

https://github.com/rainwoodman/MP-sort/pull/8

**Pull request merged**